apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        challenge = challenges[challengeId];
+    }
+
+    /**
+    * @dev Fetch an action instance along with its current challenge by a dispute identification number
+    * @param _disputeId Identification number of the dispute on the arbitrator
+    * @return actionId Identification number of the action associated with the dispute
+    * @return action Action instance associated with the dispute
+    * @return challengeId Identification number of the challenge associated with the dispute
+    * @return challenge Current challenge instance associated with the dispute
+    */
+    function _getDisputedAction(uint256 _disputeId) internal view
+        returns (
+            uint256 actionId,
+            Action storage action,
+            uint256 challengeId,
+            Challenge storage challenge
+        )
+    {
+        challengeId = challengeByDispute[_disputeId];
+        require(_existChallenge(challengeId), ERROR_DISPUTE_DOES_NOT_EXIST);
+
+        challenge = challenges[challengeId];
+        actionId = challenge.actionId;
+        action = _getAction(actionId);
+        require(action.currentChallengeId == challengeId, ERROR_DISPUTE_DOES_NOT_EXIST);
+    }
+
+    /**
+    * @dev Tell the current Agreement setting identification number
+    * @return Identification number of the current Agreement setting
+    */
+    function _getCurrentSettingId() internal view returns (uint256) {
+        return settingsLength - 1; // an initial setting is created during initialization, thus length will be always greater than 0
+    }
+
+    /**
+    * @dev Tell the information related to a signer
+    * @param _signer Address being queried
+    * @return lastSettingIdSigned Identification number of the last agreement setting signed by the signer
+    * @return mustSign Whether the requested signer needs to sign the current agreement setting before submitting an action
+    */
+    function _getSigner(address _signer) internal view returns (uint256 lastSettingIdSigned, bool mustSign) {
+        lastSettingIdSigned = lastSettingSignedBy[_signer];
+        mustSign = lastSettingIdSigned < _getCurrentSettingId();
+    }
+
+    /**
+    * @dev Tell the settings applicable for an action
+    * @param _action Action instance to query
+    * @return title String indicating a short description
+    * @return content Link to a human-readable text that describes the initial rules for the Agreements instance
+    * @return arbitrator Address of the IArbitrator that will be used to resolve disputes
+    */
+    function _getSettingFor(Action storage _action) internal view returns (string title, bytes content, IArbitrator arbitrator) {
+        uint256 settingId = _action.settingId;
+        require(settingId < settingsLength, ERROR_MISSING_AGREEMENT_SETTING);
+
+        Setting storage setting = settings[settingId];
+        title = setting.title;
+        content = setting.content;
+        arbitrator = setting.arbitrator;
```

Comment:
Maybe we could just return the storage pointer for `Setting`? It looks like all the usages of this function don't use all these variables.